### ISSUE TITLE
Single Parents for inexact rings independent of precision

### ISSUE THREAD

**Issue by simonbrandhorst** [oscar-system/Oscar.jl#1619]
The idea is to have a single parent for the elements of the same (inexact) field independent of their respective precisions.

As far as I know the elements themselves know about their respective precision, so the parent knowing the precision does not add any new value. 

The constructors would then be similar to `BigFloat` 
and take, say a rational number together with a precision. For instance
`RR(1//3,precision=100)` 

Here is an example of the current behavior over the reals. Apparently some kind of promotion happens. 
```
julia> RR1 = Nemo.RealField(23)
Real Field with 23 bits of precision and error bounds

julia> RR2 = Nemo.RealField(5)
Real Field with 5 bits of precision and error bounds

julia> RR1(1)*RR2(23)
23.00000

julia> parent(RR1(1)*RR2(23))
Real Field with 23 bits of precision and error bounds
```
Similar examples apply for p-adics and power-series rings.

**Comment by thofma** [comment]
What is the precision of the result of an operation in this model?

**Comment by simonbrandhorst** [comment]
My impression is that the element knows it?

**Comment by thofma** [comment]
The computation of `x * y` must know with which precision the result should be computed. For `BigFloat` this is the global default precision, aka the precision of the parent. This is where the precision of the parent usually comes into play.

**Comment by fieker** [comment]
On Wed, Oct 12, 2022 at 05:27:49AM -0700, Tommy Hofmann wrote:
> The computation of `x * y` must know with which precision the result should be computed. For `BigFloat` this is the global default precision, aka the precision of the parent. This is where the precision of the parent usually comes into play. 
> 
We had a diskussion about this over lunch and need to continue. My
opinion is that
 - qAdic, qAdic, Carlo-Localfield
 - PowerSeries, Laurent, Pussuix
 - ArbField, AcbField, BigFloat
Should have as uniform an interface as possible (and practical)

A suggestion is
 - singleton parent with a default precision (that can be changed)
 - operations are performed to max. prec. allowed by input
   (this requires special functions for op(exact, inexact) and such)
 - default precision is used iff an exact object is absorbed into the
   ring
 - elements also support setprecision and setprecision!

Part of a rationale:
 - the implementation allows mixing of precisions merrily
 - one gets mad if the parent check (are the rings of the same precision
   - which is completely independed of the elements having the same
   precision!) is only performed since some integer is stored in the
   ring
 - changing precision efficiently is at the core of any Hensel/Newton
   algorithms
 - handling precision in rings the "current" way is really tricky in
   towers
   (mty recent was computing logs by evaluation of the "usual" power
   series: all my "obvious" rings were at high precision, the element
   was precise, but the x^i/i the division coerced i into the prime
   field - which had low precision...
 - a uniform interface is always nice
 - I fail to see any advantage of having different ArbRing(n) around

What are your thoughts?

**Comment by tthsqe12** [comment]
> I fail to see any advantage of having different ArbRing(n) around

How can I get my pi then?

> As far as I know the elements themselves know about their respective precision

`arb`s do not know about their "precisions" except through their parents.

**Comment by fieker** [comment]
On Wed, Oct 12, 2022 at 07:15:02AM -0700, dan wrote:
> > I fail to see any advantage of having different ArbRing(n) around
> 
> How can I get my pi then?
RR(pi) - afetr you set the precsion in RR to 100:

setprecision(RR, 100) do
  p = RR(pi)
end

or setprecision!(RR, 100); p = RR(pi)

or, possibly RR(pi, prec = 100)

**Comment by tthsqe12** [comment]
So, I see the pi possibilities as
1. `setprecision(RR, 100) do; return const_pi(RR); end` ick
2. `setprecision(RR, 100); const_pi(RR)` ick
3. `const_pi(RR, 100)` ok

There are further problems with precisionless parents in arb arithmetic.
Everything in arb comes down to
``add!(x, a, b, p)``
which does
``x = ball containing a + b with intermediate precision restricted by p bits``
If p is low, the resulting ball will be of low quality (radius higher than expected), and, in any case, the complexity of the operation is supposed to be polynomial in `p`.

I have found the precision-stored-in-parent wrappers of arb to be useless in my own applications, and have just used the `add!(x,a,b,p)` stuff. If you want to support `+` with precisionless parents, each `arb_t` will have to be walking around with a `.prec` member and `+` will have to figure out a `p` from the precision of the two arguments.

**Comment by tthsqe12** [comment]
Maybe while we are at it, we can correct this apparent disaster, which is why I stayed away from the arb parents.
```
julia> using Nemo

julia> a = const_pi(ArbField(200)); b = const_pi(ArbField(100));

julia> parent(a+b), parent(b+a)
(Real Field with 200 bits of precision and error bounds, Real Field with 100 bits of precision and error bounds)
```

**Comment by thofma** [comment]
Yes, this is on 'purpose' because someone just removed the parent check to make 'mixed' things work. We are currently discussing how to proceed with the parent-less arbs. Do you know a good way to cook up a `p` from `x, y` to make `add!(arb(), x, y, p)` do something useful? Would `min(m, n)` be a useful thing where `m = arf_bits(x.midpoint)` and `n = arf_bits(y.midpoint)` (+ handling the special number case)?

Or should we tag the precision onto each `arb_t` which was used when creating it?

**Comment by tthsqe12** [comment]
The `arf_prec` thing or anything like it is probably a dead end because balls can be exact. The `.parent` member should be replaced with an Int `.prec`, and promotion should generally use the min. The downside is that now the constructors for infinity and NaN need a precision argument, which looks silly.

**Comment by thofma** [comment]
I discussed this in length with @fieker. There are two options to go parent-less:
1. Elements will be plain `arb_t`. We have a global `RR` object which has some `prec::Int` field that will be used when creating elements (a la `RR(2)`), or to determine the `slong prec` argument for the C-calls when doing `x + y` or `sin(x)`. We will provide the `set_precision(RR, p) do` syntax.
2. Elements will be `arb_t` with a `prec::Int` field tagged onto them (this will be populated when creating the element). It is clear how this will be called when doing e.g. `sin(x)`. What about `x + y`? When I mentioned taking `min(x.prec, y.prec)` @fieker wondered about the following situation. Assume `x = 1.2345` and `y = 0.00001`, so `x` with precision `5` and `y` with precision `1`. Computing `x + y` with precision `1` seems like a waste, since one will lose some information? (I did not add the radii for `x` and `y` to keep it simple). It could also all be a misunderstanding about the term "precision".

@tthsqe12: Can you give your opinion on the two proposals and comment on the question in the second one?

**Comment by tthsqe12** [comment]
1. How many global RR objects? Bye bye thread safety.
2. "Computing x + y with precision 1 seems like a waste" There are any number of precision optimizations that can be applied. For example, cos(x) gains precision when x is close to a mutliple of pi, arctan(x) gains precision as |x|->infty. As long as the `add!(x,a,b,p)` is available, we have full access to arb. If you want you, can compute `+` with `p=max(a.prec, b.prec)` or `p=(a.prec+b.prec+1)>>1`, but then if the returned real has a tagged precision of `min(x.prec, y.prec)`, it probably doesn't matter much. It's the user's fault for mixing parents.

**Comment by thofma** [comment]
I am sure we came to some conclusion, but I forgot to write it down. I think we wanted to say bye to thread safety, if I remember correctly?

**Comment by fieker** [comment]
On Fri, Nov 18, 2022 at 11:04:14AM -0800, Tommy Hofmann wrote:
> I am sure we came to some conclusion, but I forgot to write it down. I think we wanted to say bye to thread safety, if I remember correctly?
> 
Do I deterct some bias here? The precision can as "easily" be thread
local as any other parameter. But this is exactly, possibly
unfortunately, how julia does it for the BigFloat

I do not mind having may different p-adic rings - however, the precision
of any element is mostly unrelated to the ring precision.

I'm happy with any model that allows efficient lifting where elements
(and the arithmetic) needs to run at different precision - without
copying data all the time

**Comment by thofma** [comment]
Ah, sorry, no. I just used the term "bye thread safety" to refer to option number 1). I am happy with that approach.
